PlantNet plugin error 500

Investigate error 500 at 18h35


----

server side logs 

```
2020-07-12T18:35:16.156883+00:00 app[web.1]: [31m[2020-07-12 18:35:16.156] [ERROR] TwitterClient - [39mPOST statuses/update:{"in_reply_to_status_id":"1281121047589371906","status":"@LunaManaShop Pl@ntnet identifie (à 28.02%) Plumeria obtusa (fam Apocynaceae)\n\n\nfeuille - Omolla Srikanth\nhttp://tinyurl.com/ya6emsam\n\n#BetPlantnet #IndentificationDePlantes"} - err:Error: You attempted to reply to a Tweet that is deleted or not visible to you.
2020-07-12T18:35:16.157318+00:00 app[web.1]: [31m[2020-07-12 18:35:16.156] [ERROR] PlantnetBTP - [39m{ inspect: [Function: inspect] }
2020-07-12T18:35:16.157486+00:00 app[web.1]: [33m[2020-07-12 18:35:16.157] [WARN] BotEngine - [39mplugin response status:500 msg:impossible de répondre au tweet
2020-07-12T18:35:16.165090+00:00 heroku[router]: at=info method=GET path="/hook" host=botentrain.herokuapp.com request_id=8a971197-232e-45ff-b4ac-bae27c497c50 fwd="116.203.134.67" dyno=web.1 connect=1ms service=1823ms status=500 bytes=313 protocol=https
```

reproduced 16 times with effect that disable cron job. 

```
2020-07-14T18:15:34.093502+00:00 app[web.1]: 2020-07-14T18:15:34.093502+00:00 app[web.1]: Aller on joue 🌷 Vous avez l'habitude ^^, quelle est cette fleur ?2020-07-14T18:15:34.093503+00:00 app[web.1]: 2020-07-14T18:15:34.093503+00:00 app[web.1]: J'enverrai un petit bon de réduction en message privé à toutes celles/ceux qui me donneront la bonne réponse, comme la dernière fois ^^2020-07-14T18:15:34.093504+00:00 app[web.1]: 2020-07-14T18:15:34.093504+00:00 app[web.1]: A tout 😘 https://t.co/UzjeEpuSCk2020-07-14T18:15:34.319128+00:00 app[web.1]: [31m[2020-07-14 18:15:34.318] [ERROR] TwitterClient - [39mPOST statuses/update:{"in_reply_to_status_id":"1281121047589371906","status":"@LunaManaShop Pl@ntnet identifie (à 28.02%) Plumeria obtusa (fam Apocynaceae)\n\n\nfeuille - Omolla Srikanth\nhttp://tinyurl.com/ya6emsam\n\n#BetPlantnet #IndentificationDePlantes"} - err:Error: You attempted to reply to a Tweet that is deleted or not visible to you.2020-07-14T18:15:34.326777+00:00 app[web.1]: [31m[2020-07-14 18:15:34.322] [ERROR] PlantnetBTP - [39m{ inspect: [Function: inspect] }2020-07-14T18:15:34.340572+00:00 app[web.1]: [33m[2020-07-14 18:15:34.340] [WARN] BotEngine - [39mplugin response status:500 msg:impossible de répondre au tweet2020-07-14T18:15:34.343279+00:00 heroku[router]: at=info method=GET path="/hook" host=botentrain.herokuapp.com request_id=eeba94d6-2faf-41b8-9617-cb01737e3c9d fwd="116.203.134.67" dyno=web.1 connect=0ms service=2495ms status=500 bytes=313 protocol=https
```

one more - use case tweet has been deleted:

```
2020-10-08T09:35:07.408828+00:00 app[web.1]: [32m[2020-10-08 09:35:07.408] [INFO] TwitterClient - [39mPlan to reply(234) to => https://twitter.com/LunaManaShop/status/1314098280901550080
2020-10-08T09:35:07.409043+00:00 app[web.1]: [32m[2020-10-08 09:35:07.408] [INFO] TwitterClient - [39m 2020-10-08 07:00:19 --- @LunaManaShop: Coucou, le petit jeu juste comme ça, quelle est cette fleur ? 
2020-10-08T09:35:07.409044+00:00 app[web.1]: 💗🌺🌷
2020-10-08T09:35:07.409044+00:00 app[web.1]: 
2020-10-08T09:35:07.409044+00:00 app[web.1]: Juste parce que je trouvais la photo jolie :)
2020-10-08T09:35:07.409045+00:00 app[web.1]: Vous allez encore tous/tes trouver la bonne réponse :D https://t.co/FUvQc6NL8t
2020-10-08T09:35:07.594863+00:00 app[web.1]: [31m[2020-10-08 09:35:07.594] [ERROR] TwitterClient - [39mPOST statuses/update:{"in_reply_to_status_id":"1314098280901550080","status":"@LunaManaShop Bonjour, j'ai interrogé Pl@ntnet pour tenter d'identifier votre première image mais cela n'a pas donné de résultat concluant 😩 (score>20%).\nAstuce: bien cadrer la fleur ou feuille\n\n#BetPlantnet #IndentificationDePlantes"} - err:Error: You attempted to reply to a Tweet that is deleted or not visible to you.
2020-10-08T09:35:07.595156+00:00 app[web.1]: [31m[2020-10-08 09:35:07.594] [ERROR] PlantnetBTP - [39m{ inspect: [Function: inspect] }
2020-10-08T09:35:07.595252+00:00 app[web.1]: [33m[2020-10-08 09:35:07.595] [WARN] BotEngine - [39mplugin response status:500 msg:impossible de répondre au tweet
```

--- a/bet.js
+++ b/bet.js
@@ -1,4 +1,7 @@
-const BotEnTrain = require('./core/BET.js');
-
-let bet = new BotEnTrain(process.env.PORT);
-bet.run();
+try {
+    const BotEnTrain = require('./core/BET.js');
+    let bet = new BotEnTrain(process.env.PORT);
+    bet.run();
+} catch (exception) {
+    console.info("BotEnTrain Exception: ", exception);
+}

--- a/core/TwitterClient.js
+++ b/core/TwitterClient.js
@@ -4,18 +4,14 @@ const log4js = require('log4js');
 
 const TWEET_MAX_LENGTH = 280;
 const TWEET_QUERY_MAX_LENGTH = 500;
+const TWITTER_EXCLUDED_ACCOUNTS = process.env.TWITTER_EXCLUDED_ACCOUNTS || false;
 
 class TwitterClient {
   constructor() {
     this.logger = log4js.getLogger('TwitterClient');
     this.logger.setLevel('DEBUG'); // DEBUG will show api params
-    if (!process.env.APPLICATION_CONSUMER_KEY_HERE ||
-    !process.env.APPLICATION_CONSUMER_SECRET_HERE ||
-    !process.env.ACCESS_TOKEN_HERE ||
-    !process.env.ACCESS_TOKEN_SECRET_HERE
-    ) {
-        throw "TwitterClient, please setup your environment";
-    }
+    this._assumeEnvironment();
+    this._assumeExcludeQuery();
     this.twit = new Twit({
       consumer_key: process.env.APPLICATION_CONSUMER_KEY_HERE,
       consumer_secret: process.env.APPLICATION_CONSUMER_SECRET_HERE,
@@ -72,15 +68,16 @@ class TwitterClient {
 
   // search query operators : https://developer.twitter.com/en/docs/tweets/search/guides/standard-operators
   search(searchQuery, searchCount, extendedMode, cb) {
+      var twitterQuery = searchQuery + this.excludeQuery;
       // Search parameters
       // doc: https://developer.twitter.com/en/docs/tweets/rules-and-filtering/overview/standard-operators
-      if (!searchQuery || searchQuery.length > TWEET_QUERY_MAX_LENGTH) {
-        cb("Invalid searchQuery (max " + TWEET_QUERY_MAX_LENGTH + ") - searchQuery " +
-          (searchQuery ? searchQuery.length + " - " + searchQuery : "(not set)"));
+      if (!twitterQuery || twitterQuery.length > TWEET_QUERY_MAX_LENGTH) {
+        cb("Invalid query (max " + TWEET_QUERY_MAX_LENGTH + "): " +
+          (twitterQuery ? twitterQuery.length + " - " + twitterQuery : "(not set)"));
         return;
       }
       let params = {
-        q: searchQuery,
+        q: twitterQuery,
         count: searchCount,
       };
       params.result_type = 'recent'; // 'mixed';
@@ -249,6 +246,25 @@ class TwitterClient {
       return twitterDate;
     }
   }
+
+  // ~ private
+  _assumeEnvironment() {
+    if (!process.env.APPLICATION_CONSUMER_KEY_HERE ||
+    !process.env.APPLICATION_CONSUMER_SECRET_HERE ||
+    !process.env.ACCESS_TOKEN_HERE ||
+    !process.env.ACCESS_TOKEN_SECRET_HERE) {
+        throw "TwitterClient, please setup your environment";
+    }
+  }
+
+  _assumeExcludeQuery() {
+    var exclusionValidation = RegExp(/^[a-zA-Z\;]*$/);
+    if (TWITTER_EXCLUDED_ACCOUNTS && !exclusionValidation.test(TWITTER_EXCLUDED_ACCOUNTS)) {
+        throw "TwitterClient, invalid TWITTER_EXCLUDED_ACCOUNTS, expect semicolon separated twitter accounts (ex. \"elie;jeanDidier\")";
+    }
+    this.excludeQuery = TWITTER_EXCLUDED_ACCOUNTS ? TWITTER_EXCLUDED_ACCOUNTS.split(";").map((e)=>" -from:"+e).join() : "";
+  }
+
 }
 
 module.exports = TwitterClient;

